File: second-year/second-semester/MPP/mpp-proiect-repository-raulbaciulescu/travelAgency/Persistence/src/main/java/impl/database/LocationTable.java
package impl.database;

import api.Constants;
import api.Table;
import domain.Location;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.*;

public class LocationTable implements Table<Long, Location> {
    private final Connection connection;
    private Map<Constants.Db.Queries, PreparedStatement> statements;
    private static final Logger logger = LogManager.getLogger();

    public LocationTable(Connection connection) throws SQLException {
        logger.info("Initializing LocationTable");
        this.statements = new HashMap<>();
        this.connection = connection;
        initStatements();
    }

    void initStatements() throws SQLException {
        statements.put(Constants.Db.Queries.ADD,
                connection.prepareStatement("INSERT INTO location (id, name, airport) VALUES (?, ?, ?)"
                ));
        statements.put(Constants.Db.Queries.DELETE,
                connection.prepareStatement("DELETE FROM location WHERE id = ?"
                ));
        statements.put(Constants.Db.Queries.FIND_BY_ID,
                connection.prepareStatement("SELECT * FROM location WHERE id = ?"
                ));
        statements.put(Constants.Db.Queries.GET_ALL,
                connection.prepareStatement("SELECT * FROM location"
                ));
    }
    @Override
    public void add(Location location) {
        logger.traceEntry("saving location {} ", location);
        PreparedStatement statement = statements.get(Constants.Db.Queries.ADD);
        try {
            statement.setLong(1, location.getId());
            statement.setString(2, location.getName());
            statement.setString(3, location.getAirport());
            int result = statement.executeUpdate();
            logger.trace("Saved {} instances", result);
        } catch (SQLException e) {
            logger.error(e);
            System.err.println("Error DB " + e);
        }
        logger.traceExit();
    }

    @Override
    public void delete(Long o) {

    }

    @Override
    public void update(Location elem, Location newElem) {

    }

    @Override
    public Optional<Location> findById(Long id) {
        logger.traceEntry();
        List<Location> locations = new ArrayList<>();
        PreparedStatement statement = statements.get(Constants.Db.Queries.FIND_BY_ID);
        try {
            statement.setLong(1, id);
            try (ResultSet result = statement.executeQuery()) {
                while (result.next()) {
                    long id1 = result.getLong("id");
                    String name = result.getString("name");
                    String airport = result.getString("airport");
                    Location location = new Location(name, airport);
                    location.setId(id1);
                    locations.add(location);
                }
            }
        } catch (SQLException throwables) {
            logger.error(throwables);
            System.out.println("Error db" + throwables);
        }
        logger.traceExit();
        return Optional.of(locations.get(0));
    }

    @Override
    public List<Location> getAll() {
        logger.traceEntry();
        List<Location> locations = new ArrayList<>();
        PreparedStatement statement = statements.get(Constants.Db.Queries.GET_ALL);
        try {
            try (ResultSet result = statement.executeQuery()) {
                while (result.next()) {
                    long id1 = result.getLong("id");
                    String name = result.getString("name");
                    String airport = result.getString("airport");
                    Location location = new Location(name, airport);
                    location.setId(id1);
                    locations.add(location);
                }
            }
        } catch (SQLException throwables) {
            logger.error(throwables);
            System.out.println("Error db" + throwables);
        }
        logger.traceExit();
        return locations;
    }
}

File: second-year/second-semester/MPP/mpp-proiect-repository-raulbaciulescu/travelAgency2/travelAgency2/Repository/Database/LocationTable.cs
using System;
using System.Collections.Generic;
using System.Data.SQLite;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using travelAgency2.src.Domain;
using travelAgency2.Utils;

namespace travelAgency2.Repository.Database
{
    internal class LocationTable : Table<long, Location>
    {
        private SQLiteConnection connection;
        private Dictionary<Constants.Db.Queries, SQLiteCommand> statements;
        private static readonly log4net.ILog logger = log4net.LogManager.GetLogger(System.Reflection.MethodBase.GetCurrentMethod().DeclaringType);


        public LocationTable(SQLiteConnection connection)
        {
            logger.Info("Initializing LocationTable");
            this.connection = connection;
            this.statements = new Dictionary<Constants.Db.Queries, SQLiteCommand>();
            initStatements();
        }

        void initStatements()
        {

            SQLiteCommand command1 = new SQLiteCommand(connection);
            SQLiteCommand command2 = new SQLiteCommand(connection);
            SQLiteCommand command3 = new SQLiteCommand(connection);
            SQLiteCommand command4 = new SQLiteCommand(connection);
            command1.CommandText = "INSERT INTO location (id, name, airport) VALUES (@id, @name, @airport)";
            command2.CommandText = "DELETE FROM location WHERE id = @id";
            command3.CommandText = "SELECT * FROM location WHERE id = @id";
            command4.CommandText = "SELECT * FROM location";
            statements.Add(Constants.Db.Queries.ADD, command1);
            statements.Add(Constants.Db.Queries.DELETE, command2);
            statements.Add(Constants.Db.Queries.FIND_BY_ID, command3);
            statements.Add(Constants.Db.Queries.GET_ALL, command4);
        }
        public void Add(Location location)
        {
            logger.Info("enter in add location");
            SQLiteCommand statement = statements[Constants.Db.Queries.ADD];
            try
            {
                statement.Parameters.AddWithValue("@id", location.Id);
                statement.Parameters.AddWithValue("@name", location.name);
                statement.Parameters.AddWithValue("@airport", location.airport);
                int numberofrowsaffected = statement.ExecuteNonQuery();
                logger.InfoFormat("Saved {0} instances", numberofrowsaffected);
            }
            catch (Exception ex)
            {
                logger.Error(ex);
                Console.WriteLine(ex.Message);
            }
            logger.Info("exit from add location");
        }
        public void Delete(long id)
        {
            //TODO
        }

        public void Update(Location loc1, Location loc2)
        {
            //TODO
        }
        public Location FindById(long id)
        {
            logger.Info("enter in findById location");
            List<Location> locations = new List<Location>();
            SQLiteCommand statement = statements[Constants.Db.Queries.FIND_BY_ID];
            statement.Parameters.AddWithValue("@id", id);
            SQLiteDataReader reader = statement.ExecuteReader();
            if (reader.HasRows)
            {
                while (reader.Read())
                {
                    Location location = new Location((long) reader.GetInt64(2), reader.GetString(0), reader.GetString(1));
                    locations.Add(location);
                }
            }
            reader.Close();
            logger.Info("exit from findById location");
            return locations[0];
        }

        public List<Location> GetAll()
        {
            logger.Info("enter in getAll location");
            List<Location> locations = new List<Location>();
            SQLiteCommand statement = statements[Constants.Db.Queries.GET_ALL];
            SQLiteDataReader reader = statement.ExecuteReader();
            if (reader.HasRows)
            {
                while (reader.Read())
                {
                    Location location = new Location((long)reader.GetInt64(2), reader.GetString(0), reader.GetString(1));
                    locations.Add(location);
                }
            }
            reader.Close();
            logger.Info("exit from getAll location");
            return locations;
        }
    }
}
